File: android/app/src/main/java/com/paulplay/rootconnect/NativeStoragePlugin.java
package com.paulplay.rootconnect;

import android.content.SharedPreferences;
import android.content.Context;

import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "NativeStorage")
public class NativeStoragePlugin extends Plugin {
    static final String PREFS_NAME = "root_native_storage";
    static final String STORAGE_KEY = "pairedProducts";

    @PluginMethod
    public void sync(PluginCall call) {
        String products = call.getString("products");
        if (products == null) {
            call.reject("Missing products parameter");
            return;
        }

        SharedPreferences prefs = getContext().getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        prefs.edit().putString(STORAGE_KEY, products).apply();
        call.resolve();
    }
}

File: ios/App/App/NativeStoragePlugin.swift
import Capacitor

@objc(NativeStoragePlugin)
public class NativeStoragePlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "NativeStoragePlugin"
    public let jsName = "NativeStorage"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "sync", returnType: CAPPluginReturnPromise)
    ]

    static let suiteName = "group.com.paulplay.rootconnect"
    static let storageKey = "pairedProducts"

    @objc func sync(_ call: CAPPluginCall) {
        guard let products = call.getString("products") else {
            call.reject("Missing products parameter")
            return
        }

        guard let defaults = UserDefaults(suiteName: NativeStoragePlugin.suiteName) else {
            call.reject("Failed to access App Group UserDefaults")
            return
        }

        defaults.set(products, forKey: NativeStoragePlugin.storageKey)
        call.resolve()
    }
}
